### PR TITLE
Implement enhanced Faiman temperature model

### PR DESCRIPTION
**Pull request by adriesse**
- [x] Motivated by #1594
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: pvlib/temperature.py
Comment (cwhanse) [inline pvlib/temperature.py:547]: Should this constant be 298.15 to correspond with 25C (temperature at STC)?

File: pvlib/temperature.py
Comment (cwhanse) [inline pvlib/temperature.py:481]:
```suggestion
        Wind speed measured at the same height for which the wind loss
```

File: pvlib/temperature.py
Comment (cwhanse) [inline pvlib/temperature.py:503]: Prefer [unitless]

File: pvlib/temperature.py
Comment (cwhanse) [inline pvlib/temperature.py:500]: Is "effective" adding something here? `sky_view` looks to be the fraction of the sky dome "seen" by the module, so that's the usual view factor.  I wonder if the formula could be considered part of the model and the input would be `surface_tilt`, but I'll defer to the model author.

**Review by cwhanse — COMMENTED** [review]
I think we should have a test that checks the output of this function matches the output of faiman, for default parameters.

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:500]: I snuck in the "effective" because the infra-red sky is very non-isotropic and that needs to be taken into account. The formula is really just a starting point for future work, not adequately explored and validated.

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:503]: Ok.

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:547]: No, it's for converting to Kelvin.

Comment (cwhanse) [reply pvlib/temperature.py:547]: Lost me on that intent, on line 553. I think it would be more clear to just convert `temp_air` to `tempK` or whatever name you prefer.

**Comment by cwhanse** [comment]
> > I think we should have a test that checks the output of this function matches the output of faiman, for default parameters.
> 
> You mean, the test should call faiman to determine the expected value?

Yes

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py]: I agree that was not so clear.  Perhaps the rename will help.

**Comment by adriesse** [comment]
Additional reviewer(s) invited!

**Review by cwhanse — APPROVED** [review]

**Review by kanderso-nrel — COMMENTED** [review]
@adriesse I've tagged this for 0.9.4 if you're ready to add a what'snew entry for this PR

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:422]: Thanks, I didn't know what was going on there.

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:557]: I was inclined to do that initially, but felt a bit uneasy about passing the ```heat_input``` to ```poa_global```.

**Comment by adriesse** [comment]
Last chance to comment! @jsstein @mtheristis

File: docs/sphinx/source/whatsnew/v0.9.4.rst
Comment (kanderso-nrel) [inline docs/sphinx/source/whatsnew/v0.9.4.rst:63]: Cliff and I are already listed a bit farther up :)

File: pvlib/temperature.py
Comment (kanderso-nrel) [reply pvlib/temperature.py:424]: This one is still here

**Review by kanderso-nrel — APPROVED** [review]

**Comment by mtheristis** [comment]
@adriesse, I am sorry for the delay. 

Out of curiosity, is there a reason we are only focusing on Faiman here?  Or will the others (SAPM, NOCT, PVSyst) come later?

**Comment by adriesse** [comment]
> @adriesse, I am sorry for the delay.
> 
> Out of curiosity, is there a reason we are only focusing on Faiman here? Or will the others (SAPM, NOCT, PVSyst) come later?

Good questions.  There were multiple considerations in my head.  I decided to do one in order to keep the PR to a reasonable size and work through integration issues on one function rather than four at a time.  I picked Faiman because it is the simplest but also because it is used in IEC 61853.  When this feature is released I will point it out to interested parties in the working group for that standard.

If there is user demand I see no problem to implement the other three and maybe enhance the generic linear model too.

File: docs/sphinx/source/whatsnew/v0.9.4.rst
Comment (adriesse) [reply docs/sphinx/source/whatsnew/v0.9.4.rst:63]: I guess that must have happened in the merge.

**Comment by mtheristis** [comment]
> > @adriesse, I am sorry for the delay.
> > Out of curiosity, is there a reason we are only focusing on Faiman here? Or will the others (SAPM, NOCT, PVSyst) come later?
> 
> Good questions. There were multiple considerations in my head. I decided to do one in order to keep the PR to a reasonable size and work through integration issues on one function rather than four at a time. I picked Faiman because it is the simplest but also because it is used in IEC 61853. When this feature is released I will point it out to interested parties in the working group for that standard.
> 
> If there is user demand I see no problem to implement the other three and maybe enhance the generic linear model too.

Understood.

I am just thinking that we mind end up with X versions of a same base model, and the users might be confused with their differences. Is there a way to have an option when running faiman to say "include_radiative_losses = True/False"? Or is this bad practice?

**Comment by adriesse** [comment]
> I am just thinking that we mind end up with X versions of a same base model, and the users might be confused with their differences. 

Well, it has taken a lot of years for the first enhancement to arrive and I'm not planning another one!  Presenting this as an enhancement of something that people feel familiar and comfortable with is a strategy to gain acceptance more readily.  It could
easily have been packaged as a single new model.

> Is there a way to have an option when running faiman to say "include_radiative_losses = True/False"? Or is this bad practice?

Perhaps in the system class or model chain something like this is possible.  I'm not sure.

File: pvlib/temperature.py
Comment (AdamRJensen) [inline pvlib/temperature.py:447]:
```suggestion
    '''  # noqa: E501
```

File: pvlib/temperature.py
Comment (AdamRJensen) [inline pvlib/temperature.py:528]:
```suggestion

    See Also
    --------
    pvlib.temperature.faiman

```

Comment (AdamRJensen) [reply pvlib/temperature.py:528]: I would add a similar See Also section to the original Faiman function. For the original Faiman function I would probably also refer to the other module temperature functions.

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:527]: Yes, I'll add some warning about the values.  We had some discussion already about this.  The current defaults are there to fall back to the standard Faiman model.  My second favorite option would be to have no defaults.

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:528]: I will add the Faiman reference for sure.  If you would like all temperature models to point to each other that would make a lot of changes, both now and in the future. But it could be done...

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:553]: Actually this is a different constant, but it is also in scipy so I could import it.  I recall some discussion about such imports in the past but don't recall the conclusion.

Comment (adriesse) [reply pvlib/temperature.py:553]: ```temp_air``` could be a scalar or list

File: pvlib/tests/test_temperature.py
Comment (adriesse) [reply pvlib/tests/test_temperature.py:159]: It's a possibility, but what kind of error would it catch that the others don't?

**Review by adriesse — COMMENTED** [review]
Thanks for reviewing, Adam.  I'm glad you found the work interesting!  
I responded with some comments and will hold off on code changes to see what other comments might appear...

PS. Sorry, I thought these would appear as replies to your comments.  :(

File: pvlib/temperature.py
Comment (AdamRJensen) [reply pvlib/temperature.py:527]: I would be ok with no defaults. But still, it would need a note saying that the coefficients are note the same when you add the long-wave irradiance.

File: pvlib/temperature.py
Comment (kanderso-nrel) [reply pvlib/temperature.py:528]: > all temperature models to point to each other that would make a lot of changes, both now and in the future

Proliferation of "see also" has crossed my mind as well, in the context of `pvlib.iam`.  Nowadays I lean towards only referencing functions that are more closely related than just being in the business of modeling the same phenomenon.

File: pvlib/temperature.py
Comment (AdamRJensen) [reply pvlib/temperature.py:553]: Ahh I see, the dear Boltzmann was a productive man. Though, now I might perhaps suggest just calling it sigma, which is the common symbol used to represent it. This [PR](https://github.com/pvlib/pvlib-python/pull/1083/files#diff-5b6b0d8a6e8ee2af1716d84ea1922d3c0e8c9f431024c1626117b87a91f8fc87R755) also refers to it as sigma. As a side note, *k* is typically used for the Boltzmann constant:

image

File: pvlib/temperature.py
Comment (AdamRJensen) [reply pvlib/temperature.py:528]: Fair point. Let's just keep it at the suggested edit then.

File: pvlib/temperature.py
Comment (AdamRJensen) [reply pvlib/temperature.py:553]: But couldn't the remaining input parameters then also be a list, e.g., ``poa_global``?

Comment (cwhanse) [reply pvlib/temperature.py:553]: As a general rule, I don't think it is worth the effort to accommodate type `list` as input. Scalar, numpy.array and Series should be acceptable.

Comment (cwhanse) [reply pvlib/temperature.py:553]: The function wouldn't return the same type, and we may unintentionally create an expectation that other functions would also support `list` inputs.

It's not a blocker here for me.

File: pvlib/temperature.py
Comment (AdamRJensen) [inline pvlib/temperature.py:571]:
```suggestion
    poa_global = np.asanyarray(poa_global)
    temp_air = np.asanyarray(temp_air)
    wind_speed = np.asanyarray(wind_speed)
    ir_down = np.asanyarray(ir_down)
    u0 = np.asanyarray(u0)
    u1 = np.asanyarray(u1)
    sky_view = np.asanyarray(sky_view)
    emissivity = np.asanyarray(emissivity)

    abs_zero = -273.15
    sigma = scipy.constants.Stefan_Boltzmann

    if ir_down is None:
        qrad_sky = 0.0
```

I see that the usage of np.asanyarray is very carefully chosen such that all inputs can be passed as lists and directly or indirectly be converted to arrays. The indirect conversion to arrays (by multiplying a list something array-like) seems undesirable and unintuitive to the user. Why not instead just convert all the inputs to array-like in the beginning and be straightforward about it?

**Review by AdamRJensen — APPROVED** [review]

File: pvlib/temperature.py
Comment (adriesse) [reply pvlib/temperature.py:571]: On the whole I believe this function behaves like numpy would.  The reason for the indirect and partial approach is to allow pandas and xarray data structures to be propagated from the first four arguments.  Try it, you might like it!

Comment (wholmgren) [reply pvlib/temperature.py:571]: What about this function requires any of the `np.asanyarray` calls?

Comment (wholmgren) [reply pvlib/temperature.py:571]: Ah, now I see the hidden conversation about allowing `list` input. That is not needed and comes at a cost of disallowing arbitrary array-like input that doesn't fully support the asanyarray machinery (e.g. dask, or whatever the pydata library of the month is).

Comment (adriesse) [reply pvlib/temperature.py:571]: Could you elaborate on what doesn't work here?

Comment (adriesse) [reply pvlib/temperature.py:571]: Overall I'm getting more push-back than expected on something which I though was a very simple technique to make the function very flexible.  If there is a concrete down-side I'd like to learn about it, especially the risks associated with ```asanyarray``` (which could be replaced with ```asarray``` in this particular function).  I've use this technique before and am likely to use it again unless there really is a problem with it.

Comment (wholmgren) [reply pvlib/temperature.py:571]: The function is already very flexible because you did a good job writing it :). It doesn't need `asanyarray`! Here's a concrete downside for `asanyarray`:

```python
In [1]: import numpy as np

In [2]: import dask.array as da

In [3]: da.from_array([0, 1])
Out[3]: dask.array<array, shape=(2,), dtype=int32, chunksize=(2,), chunktype=numpy.ndarray>

In [4]: np.asanyarray(da.from_array([0, 1]))
Out[4]: array([0, 1])

In [5]: np.asarray(da.from_array([0, 1]))
Out[5]: array([0, 1])
```

Comment (adriesse) [reply pvlib/temperature.py:571]: > `np.asanyarray` is not robust.

I am curious to learn more about this...

**Comment by adriesse** [comment]
Thank for all the reviews!!

**Pull request merged**
